bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/ExtendedDiscoveryServiceTracker.java

```diff
@@ -0,0 +1,155 @@
+import org.eclipse.smarthome.config.discovery.ExtendedDiscoveryService;
+import org.eclipse.smarthome.config.discovery.inbox.Inbox;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingRegistry;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link ExtendedDiscoveryServiceTracker} adds the {@link DiscoveryServiceCallback} to all
+ * {@link ExtendedDiscoveryService} instances.
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true)
+public class ExtendedDiscoveryServiceTracker {
+
+    private @Nullable Inbox inbox;
+    private @Nullable ThingRegistry thingRegistry;
+    private @Nullable DiscoveryServiceCallback discoveryServiceCallback;
+
+    private final List<ExtendedDiscoveryService> extendedDiscoveryServices = new ArrayList<>();
+    private final AtomicBoolean active = new AtomicBoolean(false);
+
+    @Reference
+    protected void addDiscoveryService(DiscoveryService discoveryService) {
+        if (discoveryService instanceof ExtendedDiscoveryService) {
+            ExtendedDiscoveryService extendedDiscoveryService = (ExtendedDiscoveryService) discoveryService;
+            extendedDiscoveryServices.add(extendedDiscoveryService);
+
+            if (active.get()) {
+                DiscoveryServiceCallback dsc = this.discoveryServiceCallback;
+                if (dsc != null) {
+                    extendedDiscoveryService.setDiscoveryServiceCallback(dsc);
+                }
+            }
+        }
+    }
+
+    protected void removeDiscoveryService(DiscoveryService discoveryService) {
+        if (discoveryService instanceof ExtendedDiscoveryService) {
+            ExtendedDiscoveryService extendedDiscoveryService = (ExtendedDiscoveryService) discoveryService;
+            extendedDiscoveryServices.remove(extendedDiscoveryService);
+        }
+    }
+
+    @Reference
+    protected void setInbox(Inbox inbox) {
+        this.inbox = inbox;
+    }
+
+    protected void unsetInbox(Inbox inbox) {
+        this.inbox = null;
+    }
+
+    @Reference
+    protected void setThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = thingRegistry;
+    }
+
+    protected void unsetThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = null;
+    }
+
+    @Activate
+    protected void activate() {
+        DiscoveryServiceCallback dsc = createDiscoveryServiceCallback();
+        this.discoveryServiceCallback = dsc;
+        active.set(true);
+
+        for (ExtendedDiscoveryService extendedDiscoveryService : extendedDiscoveryServices) {
+            extendedDiscoveryService.setDiscoveryServiceCallback(dsc);
```

Comment:
This means that the callback is now a dynamic reference for the service, i.e. it can (or cannot) come at any time during runtime, while before it was guaranteed to be be at least called immediately after being injected into the registry.
Also note that this call was previously done in a safeCaller and now directly.